T++ Redux: Preview of TPP v2.0

mattavos opened the pull request:
_Thanks to Will, Aidan, Arun, and Sam for chats, which have greatly informed the choices here!_

* Basically a full re-write with production-worthy code, tests, etc.
* Fully backwards compatible. Stuff is deprecated but not removed. The hope is that things can be removed in v2.0.
* `Expects` is deprecated. You should now use a regular `[]tpp.Expect` and Expectorise it as so:
```go
expectFoo: []tpp.Expect{
	tpp.Given(xxx).Return(yyy),
	tpp.Given(yyy).Return(zzz),
}

// ...

tpp.ExpectoriseMulti(tt.expectFoo, func() tpp.MockCall {
	return m.EXPECT().DoThing(tpp.Arg(), tpp.Arg())
})
```
* There's now `Return` which you can use to specify the exact returns:
```go
getFoo: tpp.Return(xxx, yyy)
```
* You can now assign arguments to mock calls with `Given` as so:
```go
getFoo:  tpp.Given(xxx).Return(yyy)
```
* You can now specify the number of times an expectation should be called:
```go
getFoo:  tpp.Given(xxx).Return(yyy).Once()
getBar:  tpp.Given(xxx).Return(yyy).Times(42)
```
* You can now pass default args into Expectorise, which get resolved only if no returns are specified:
```go
e.Expectorise(c, tpp.WithDefaultReturns(123, errTest))
```
* Because refactoring becomes annoying without compiler support for mock calls, you now get much more helpful errors when mocks are called with the wrong returns:
Before:
```
|| --- FAIL: TestExpectWithMockeryMockIntyThing (0.00s)
||     --- FAIL: TestExpectWithMockeryMockIntyThing/Return()_sets_up_return (0.00s)
|| panic: toReflectValues failed to transform return values: mismatched number of args: expected 2 but got 3 [recovered]
|| 	panic: toReflectValues failed to transform return values: mismatched number of args: expected 2 but got 3
```
After:
```
|| --- FAIL: TestExpectWithMockeryMockIntyThing (0.00s)
||     --- FAIL: TestExpectWithMockeryMockIntyThing/Return()_sets_up_return (0.00s)
|| panic: 
|| Return() called with the wrong arguments!
||     Function: <*testdata.MockIntyThing_DoThing_Call Value>
||     Expected: (int, error)
||     Received: (int, string, *errors.fundamental)
```

mattavos committed:
Horrible initial commit of new stuff

Works, tested against iona, but is grotesque

mattavos committed:
Split out deprecated stuff

mattavos committed:
Doc styling

mattavos committed:
lmao

mattavos committed:
Separate out reflect stuff. Tidy. Rearrange

mattavos committed:
Commenting, tidying, todo-ing

mattavos committed:
Tests for "bare" (unwrapped) mock

mattavos committed:
Much more mature []Expect handling

mattavos committed:
Woopsie

mattavos committed:
Some tests for ExpectMulti

mattavos committed:
More tests

mattavos committed:
Mockery for tests

mattavos committed:
Tests for mockery call

mattavos committed:
ExpectMulti tests for mockery call

mattavos committed:
Some reflect tests

mattavos committed:
Reflect tests

mattavos committed:
Comments

mattavos committed:
Tidying

mattavos committed:
WithDefaultReturns plus tests

mattavos committed:
Typo

mattavos committed:
Change comment on OK

Does need more thought though... Maybe Return *shouldn't* lead to the
error being automatically filled in...? I think that's right.

mattavos committed:
Update README a bit

mattavos committed:
Mark stuff deprecated

mattavos committed:
Fix the case where there's a default return and an error

mattavos committed:
Nicer error panics

mattavos committed:
Tests for structy things

Also contains a fix for returning nil structs. How did I miss that one
back in the day? :O

mattavos committed:
ExpectMulti tests for structy thing

mattavos committed:
Introduce .Arg() for placeholders set in the meta-test

mattavos committed:
Check args before potentially panicking when toReflectValuesing

mattavos committed:
For backwards compatibility, use Expect.exactReturn

See TODO comment in tpp.go

mattavos committed:
Update docs on preferred approach

mattavos commented on deprecated.go:
This just pulls out deprecated stuff and marks it as so. All this is for the chop in 2.0

mattavos committed:
Typo

mattavos committed:
Replace tpp.Arg() with mock.Anything for empty mocks

aidanavos reviewed (approved):
I've only glanced through the code, but having used the new functionality a bit in my branch I'd say this is a marked improvement! Great work @mattavos, t++ is shaping up nicely!

aidanavos commented:
Another interesting little find: A case of t++ not being able to use `nil` in a place I'm pretty sure it should.

In this case:
```
	}, {
		name:           "ERR: publisher encounters store error",
		withReqMachine: d.AsMachineSet(machineId),
		expectGetMachineOverview: []tpp.Expect{
			tpp.Given(d.AsMachineSet(machineId)).Return(nil, errTest).Once(),
		},
		wantResp_0: &want{err: true},
	}, {
```

with this body code:
```
	tpp.ExpectoriseMulti(tt.expectGetMachineOverview, func() tpp.MockCall {
		return mStore.EXPECT().GetMachineInfos(tpp.Arg())
	})
```

For this underlying function:
```
type IStore interface {
	GetMachineInfos(d.MachineSet) ([]d.MachineOverview, error)
	...
}
```

I'm getting a reflection error:
```
panic: toReflectValues failed to transform return values: cannot handle nil for non-interface or non-pointer type: []domain.MachineOverview [recovered]
        panic: toReflectValues failed to transform return values: cannot handle nil for non-interface or non-pointer type: []domain.MachineOverview
```

Which looks wrong/unexpected to me, as `nil` is a perfectly valid value for a slice. Replacing `nil` with `[]d.MachineOverview{}` gives no error and the test passes.

Any thoughts on whether this is right/expected or not, or if I've missed something obvious?

aidanavos commented on reflect.go:
```diff
@@ -0,0 +1,322 @@
+		if arg != nil {
+			values[i] = reflect.ValueOf(arg)
+		} else {
+			switch argType.Kind() {
+			case reflect.Interface, reflect.Ptr:
+				values[i] = reflect.Zero(argType)
+			default:
+				return nil, fmt.Errorf(
+					"cannot handle nil for non-interface or non-pointer type: %s",
+					argType,
+				)
+			}
```
```suggestion
			switch argType.Kind() {
			case reflect.Interface, reflect.Ptr, reflect.Slice:
				values[i] = reflect.Zero(argType)
			default:
				return nil, fmt.Errorf(
					"cannot handle nil for non-interface, non-pointer, non-slice type: %s",
					argType,
				)
			}
```
Something like this would fix the problem I described in my comment above (I reran the test and it passed) but I have no idea if this would be logically incorrect for some other reason.

mattavos replied:
oooooooo

mattavos committed:
Fix nil type reflection

mattavos committed:
Generic tests for Expect

mattavos committed:
Tidy

mattavos committed:
Rewrite tests in a generic way to avoid repetition

mattavos committed:
Loads more tests

mattavos committed:
Don't export the new fields

The aim is to ultimately make all Expect data fields unexported. So
don't export the new ones!

mattavos committed:
Get rid of dependency on slices pkg (only exists in newer go)

mattavos committed:
Test for mock.MatchedBy
